fluent-langneg/src/accepted_languages.js

```diff
@@ -1,7 +1,26 @@
-export default function acceptedLanguages(string = "") {
-  if (typeof string !== "string") {
+function parseAcceptLanguageEntry(entry, index) {
+  const langWithQ = entry.split(";").map(u => u.trim());
+  let q = 1.0;
+  if (langWithQ.length > 1) {
+    const qVal = langWithQ[1].split("=").map(u => u.trim());
+    if (qVal.length === 2 && qVal[0].toLowerCase() === "q") {
+      const qn = Number(qVal[1]);
+      q = isNaN(qn) ? 0.0 : qn;
+    }
+  }
+  return { index: index, lang: langWithQ[0], q };
+}
+
+export default function acceptedLanguages(acceptLanguageHeader = "") {
+  if (typeof acceptLanguageHeader !== "string") {
     throw new TypeError("Argument must be a string");
   }
-  const tokens = string.split(",").map(t => t.trim());
-  return tokens.filter(t => t !== "").map(t => t.split(";")[0]);
+  const tokens = acceptLanguageHeader.split(",").map(t => t.trim())
+    .filter(t => t !== "");
+  const langsWithQ = [];
+  tokens.forEach((t, index) =>
+    langsWithQ.push(parseAcceptLanguageEntry(t, index)));
+  // order by q descending, keeping the header order for equal weights
+  langsWithQ.sort((a, b) => a.q === b.q ? a.index - b.index : b.q - a.q);
```

Comment:
Non-blocking cleanup suggestion:

```js
function parseAcceptLanguageEntry(entry) {
  const langWithQ = entry.split(";").map(u => u.trim());
  let q = 1.0;
  if (langWithQ.length > 1) {
    const qVal = langWithQ[1].split("=").map(u => u.trim());
    if (qVal.length === 2 && qVal[0].toLowerCase() === "q") {
      const qn = Number(qVal[1]);
      q = isNaN(qn) ? 0.0 : qn;
    }
  }
  return { lang: langWithQ[0], q };
}

const langsWithQ = Array.from(tokens.map(parseAcceptLanguageEntry).entries());
// order by q descending, keeping the header order for equal weights
langsWithQ.sort(([aidx, aval], [bidx, bval]) => aval.q === bval.q ? aidx - bidx : bval.q - aval.q);
return langsWithQ.map(([idx, val]) => val.lang);
```

I hand-written it so it may not launch but I hope the intention is expressed in my snippet. I'm trying to save us from reallocating the array in the loop. Alternatively, you could also likely do `const langsWithQ = Array(tokens.length);` which I guess is simpler.
The other thing I wanted to improve in my snippet is to not carry around the index, since the `langsWithQ` preserves the order of `tokens`, so there's no value in storing it on the `langWithQ` struct.

@blushingpenguin - do you like the proposed changes or would you prefer to land as-is?

Reply:
Nicer your way I think -- I wasn't aware of Array.entries.  I have updated the code and retested it.